traffic_ops/traffic_ops_golang/deliveryservice/consistenthash/consistenthash.go

```diff
@@ -0,0 +1,134 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import (
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/api"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+// struct for the response object from Traffic Router
+type TRConsistentHashResult struct {
+	ResultingPathToConsistentHash string `json:"resultingPathToConsistentHash"`
+	ConsistentHashRegex           string `json:"consistentHashRegex"`
+	RequestPath                   string `json:"requestPath"`
+}
+
+// struct for the incoming request object
+type TRConsistentHashRequest struct {
+	ConsistentHashRegex string `json:"regex"`
+	RequestPath         string `json:"requestPath"`
+	CdnID               int64  `json:"cdnId"`
+}
+
+// endpoint to test Traffic Router's Pattern-Based Consistent Hashing feature
+// /api/1.5/consistenthash
+func Post(w http.ResponseWriter, r *http.Request) {
+	inf, userErr, sysErr, errCode := api.NewInfo(r, nil, nil)
+	if userErr != nil || sysErr != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, errCode, userErr, sysErr)
+		return
+	}
+	defer inf.Close()
+
+	req := TRConsistentHashRequest{}
+	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusBadRequest, errors.New("malformed JSON: "+err.Error()), nil)
+		return
+	}
+
+	responseFromTR, err := getPatternBasedConsistentHash(inf.Tx.Tx, req.ConsistentHashRegex, req.RequestPath, req.CdnID)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, errors.New("getting pattern based consistent hash from Traffic Router: "+err.Error()))
+		return
+	}
+
+	consistentHashResult := TRConsistentHashResult{}
+	json.Unmarshal(responseFromTR, &consistentHashResult)
+
+	api.WriteResp(w, r, consistentHashResult)
+}
+
+const RouterRequestTimeout = time.Second * 10
+
+// queries database for active Traffic Router on the CDN specified by cdnId
+// passes regex and requestPath to the Traffic Router via API request,
+// and returns the response
+func getPatternBasedConsistentHash(tx *sql.Tx, regex string, requestPath string, cdnId int64) ([]byte, error) {
+	q := `
+SELECT concat(server.host_name, '.', server.domain_name) AS fqdn,
+   parameter.value AS apiport
+  FROM (((server
+    JOIN profile ON ((profile.id = server.profile)))
+    JOIN profile_parameter ON ((profile_parameter.profile = profile.id)))
+    JOIN parameter ON ((parameter.id = profile_parameter.parameter)))
+    JOIN status ON ((server.status = status.id))
+ WHERE ((server.type = (select id from type where name = 'CCR')) AND
+ (parameter.name = 'api.port'::text) AND
+ (status.name = 'ONLINE') AND
+ (server.cdn_id = $1))
+`
+	rows, err := tx.Query(q, cdnId)
+	defer rows.Close()
```

Comment:
I think the `defer` has to come after the `if err != nil` check.

Reply:
Good catch. Even if it's technically permissible, it's always safer to treat all other values as undefined, if an err is non-nil

Reply:
actually, since we really only need one random traffic router to try, I think we should add this to the query:
```
ORDER BY RANDOM()
LIMIT 1
```
and use `tx.QueryRow()` instead of `tx.Query()` like:
```
if err := tx.QueryRow(q, cdnId).Scan(&trafficRouter); err != nil {
...
}
```

Reply:
using `tx.QueryRow()` makes the code much cleaner